### ISSUE TITLE
Can I use example packing and train on completions only at the same time using SFTTrainer?

### ISSUE THREAD

**Issue by salahzoubi** [huggingface/trl#772]
I'm trying to maximize training efficiency by using example packing and train on completions only and not the entire context. Is there a way to do this? From the documentation on the SFTTrainer it seems like you can only use one or the other, but I'm wondering if I could do both at the same time? 

Let's say my data looks something like this "### Instruction: instructions ### Input: input ### Response: response" if I use a data collator on a packed example, it'll probably take everything after ### Response including the next samples. Not sure if this is possible? If so, any particular resources on how to do this?

**Comment by younesbelkada** [comment]
Hi @salahzoubi 
Thanks for your message, from what I have understood, I think what you are referring to is to train on the completions only https://huggingface.co/docs/trl/sft_trainer#train-on-completions-only using `DataCollatorForCompletionOnlyLM` please let me know if this helps

**Comment by salahzoubi** [comment]
Hello @younesbelkada, 

so from my understanding, `DataCollatorForCompletionOnlyLM` takes a raw example, and splits it into inputs and labels, right? My question is, can this be done for a `ConstantPackedLengthDataset`? In other words, let's say I have a sequence_length of 4096 and I pack 4x1024 examples into one example, is there a way to apply `DataCollatorForCompletionOnlyLM` on that example so that it splits that 4x1024 example in 4 inputs and 4 labels accordingly? 

From what I'm seeing right now, the data collator only does this for a single example and not multiple examples that are packed into one.

**Comment by younesbelkada** [comment]
Hi @salahzoubi 

Thanks a lot for your reply, I think that I see your point now, in case of packed dataset I think that `DataCollatorForCompletionOnlyLM` should also work, check for instance this test on main: https://github.com/huggingface/trl/blob/0a6c42c12c637bb7f28782fa72ec45dd64bce0bd/tests/test_sft_trainer.py#L470-L500 - however returning 4 independent samples from a single concatenated sample is currently not supported. If you pack 4 sentences into one, you will get a single `input_ids` and a single `attention_mask` but the mask being properly created

**Comment by younesbelkada** [comment]
Closing for now, let me know if you still have more questions
